rust/src/client/mobile_client/client.rs

```diff
@@ -0,0 +1,364 @@
+            proxy,
+            Participant::<Undefined>::new(participant_settings.into()),
+            local_model,
+            global_model,
+        )
+        .into())
+    }
+
+    pub async fn next(self) -> Self {
+        match self {
+            ClientStateMachine::Undefined(state) => state.next().await,
+            ClientStateMachine::Sum(state) => state.next().await,
+            ClientStateMachine::Update(state) => state.next().await,
+            ClientStateMachine::Sum2(state) => state.next().await,
+        }
+    }
+}
+
+pub struct ClientState<Type> {
+    proxy: Proxy,
+    round_params: RoundParameters,
+    participant: Participant<Type>,
+    local_model: Rc<RefCell<Option<Model>>>,
+    global_model: Rc<RefCell<Option<Model>>>,
+}
+
+impl<Type> ClientState<Type> {
+    async fn check_round_freshness(&self) -> Result<(), ClientError> {
+        debug!("fetching round parameters");
+        let round_params = self.proxy.get_round_params().await?;
+        if round_params.seed != self.round_params.seed {
+            info!("new round parameters");
+            Err(ClientError::RoundOutdated)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn reset(self) -> ClientState<Undefined> {
+        warn!("reset client");
+        ClientState::<Undefined>::new(
+            self.proxy,
+            self.participant.reset(),
+            self.local_model,
+            self.global_model,
+        )
+    }
+}
+
+impl ClientState<Undefined> {
+    fn new(
+        proxy: Proxy,
+        participant: Participant<Undefined>,
+        local_model: Rc<RefCell<Option<Model>>>,
+        global_model: Rc<RefCell<Option<Model>>>,
+    ) -> Self {
+        Self {
+            proxy,
+            round_params: RoundParameters::default(),
+            participant,
+            local_model,
+            global_model,
+        }
+    }
+
+    async fn next(mut self) -> ClientStateMachine {
+        info!("new participant with undefined task");
+        if let Err(err) = self.fetch_round_params().await {
+            error!("{:?}", err);
+            return self.reset().into();
+        };
+
+        let Self {
+            proxy,
+            round_params,
+            participant,
+            local_model,
+            global_model,
+        } = self;
+
+        let participant_type = participant.determine_type(
+            round_params.seed.as_slice(),
+            round_params.sum,
+            round_params.update,
+        );
+
+        match participant_type {
+            Type::Unselected(unsel_par) => {
+                info!("unselected");
+                ClientState::<Undefined>::new(proxy, unsel_par.reset(), local_model, global_model)
+                    .into()
+            }
+            Type::Summer(sum_par) => {
+                ClientState::<Sum>::new(proxy, round_params, sum_par, local_model, global_model)
+                    .into()
+            }
+            Type::Updater(upt_pat) => {
```

Comment:
```suggestion
            Type::Updater(upt_par) => {
```
or maybe just `Type::...(participant)` for all three bindings.